blogposts/2021/optimizing-a-code-intelligence-commit-graph.md

```diff
@@ -0,0 +1,699 @@
+
+In this case, it was **very** noticeable.
+
+![critical db size](https://sourcegraphstatic.com/blog/commit-graph-optimizations/full-disk.png)
+
+## Pick 3: low latency, low memory usage, low disk usage, and developer sanity
+
+Home stretch - we can do this. To summarize:
+
+1. The speed at which we produce the commit graph is no longer a problem, and
+1. The resources we require to produce the commit graph is no longer a problem, but
+1. The amount of data we're writing into Postgres _is_ a problem.
+
+Our [final and successful effort](https://github.com/sourcegraph/sourcegraph/pull/16636) to fix these time and space issues attacked the problem in two parts.
+
+First, we've changed the commit graph traversal behavior to look only at ancestors. Since we started to ignore the bulk of the commit graph that existed before the first commit with LSIF data, looking into the _future_ for LSIF data now has limited applicability. Generally, users will be on the tip of a branch (either the default branch, or a feature branch if reviewing a pull request). We can still answer code intelligence queries for these commits, as they necessarily occur after _some_ LSIF data has been uploaded. Additionally, an unrelated [bug fix](https://github.com/sourcegraph/sourcegraph/pull/16733) caused descendant-direction traversals to increase memory usage. So at this point, it just seemed sensible to stop worrying about this reverse case, which was never truly necessary from a usability standpoint. Looking in one direction reduces the amount of data we need to store by about 50%.
+
+Second, we re-applied our tricks described in the previous section. We had good luck at throwing out huge amounts of data which we could efficiently recalculate when necessary, and save the resources that would otherwise be required to store them. This concept is known as [rematerialization](https://en.wikipedia.org/wiki/Rematerialization) in the compiler ecosystem, where values may be computed multiple times instead of storing and loading the already-computed value from memory in the case where a load/store pair is more expensive than the computation itself, or if it would otherwise increase register pressure.
+
+Instead of writing the set of visible uploads per commit, what if we only store the visible uploads for commits that can't trivially recomputed? We've already determined the set of commits that can be easily rematerialized - we can just move the rematerialization from database insertion time to query time.
```

Comment:
```suggestion
Instead of writing the set of visible uploads per commit, what if we only store the visible uploads for commits that can't trivially recomputed? We've already determined the set of commits that can be easily rematerialized—we can just move the rematerialization from database insertion time to query time.
```